### PR TITLE
Error handling

### PR DESCRIPTION
**Pull request by elena-kolevska**
# Description

Improves developer experience on error handling for the new Rich gRPC Errors from Dapr

## Issue reference

https://github.com/dapr/python-sdk/issues/648

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests
* [X] Extended the documentation

### CODE REVIEW

**Comment by elena-kolevska** [comment]
@berndverst pls let me know if the DaprGrpcError design and naming look ok to you. I would also like to get your opinion on the integration testing approach I took. I think it can be helpful for other cases too.

**Comment by berndverst** [comment]
LGTM. Can you resolve the conflict now that I merged the ruff stuff :)

**Comment by elena-kolevska** [comment]
Marking this ready so the code can be reviewed while I'm working on examples and docs.

**Comment by elena-kolevska** [comment]
Ok, examples and docs are ready :)

File: tox.ini
Comment (berndverst) [inline tox.ini:48]: That's fine - but in the future just use a `requirements.txt` for your specific example you are adding.

File: tox.ini
Comment (berndverst) [inline tox.ini:50]: You need to actually register the `error_handling` folder here as well, so we can run this via `tox -e examples`.

File: examples/error_handling/error_handling.py
Comment (berndverst) [inline examples/error_handling/error_handling.py:17]: To avoid weird log delays which causes the mechanical markdown test to fail, use `print("SOMETHING", flush=True)`

For all of these log lines.

File: examples/error_handling/error_handling.py
Comment (berndverst) [inline examples/error_handling/error_handling.py:16]: Could you also add a nice `json()` method for `DaprGrpcError` which tries to give you a JSON version of all of the properties etc?

File: dapr/clients/exceptions.py
Comment (berndverst) [inline dapr/clients/exceptions.py:47]: Please add a `json()` method which can print a map of all the error attributes and their values.

That would be most convenient for just dumping the entire error in a more readable manner.

**Review by berndverst — CHANGES_REQUESTED** [review]

File: examples/error_handling/error_handling.py
Comment (elena-kolevska) [reply examples/error_handling/error_handling.py:16]: Great idea! On it!

**Comment by elena-kolevska** [comment]
I addressed the review comments and I simplified the DaprGrpcError structure. 🙏

File: tox.ini
Comment (berndverst) [inline tox.ini:49]:
```suggestion
    ./validate.sh metadata
    ./validate.sh error_handling
```

Comment (berndverst) [reply tox.ini:49]: You had resolved my comment here without actually doing the thing I told you to do :) Otherwise you would have seen that this test is failing!

File: examples/error_handling/error_handling.py
Comment (berndverst) [inline examples/error_handling/error_handling.py:18]:
```suggestion
        print(f'Message: {err.details()}', flush=True)
```

File: examples/error_handling/error_handling.py
Comment (berndverst) [inline examples/error_handling/error_handling.py:1]: This example is currently failing.

File: examples/error_handling/README.md
Comment (berndverst) [inline examples/error_handling/README.md:36]: You might have to switch to regex mode or substring match mode (supported by mechanical markdown) if you find you don't reliably get the exact same output

File: setup.cfg
Comment (berndverst) [inline setup.cfg:30]:
```suggestion
    grpcio-status>=1.37.0
```

Comment (berndverst) [reply setup.cfg:30]: It's important we use a low version here. Otherwise this forces all other users of the Dapr Python SDK to use this version with other tools that require it. And worse, grpcio-status 1.60.0 requires grpcio 1.60.0. And we cannot be totally sure that this is ok for every user of the Dapr Python SDK.

**Review by berndverst — CHANGES_REQUESTED** [review]

**Comment by berndverst** [comment]
I now updated the tox configuration to include your error handling test app. I am expected it to fail (it fails for me locally). You'll have to debug that one some more.

**Comment by berndverst** [comment]
If this PR requires Dapr 1.13, please let me know. In that case we will not be able to successfully run the CI tests until we have Dapr 1.13.0rc1 tagged

**Review by berndverst — APPROVED** [review]

**Comment by berndverst** [comment]
I will rerun tests and merge this when 1.13.0rc1 has been tagged. At that point this should all pass :) (I tried it locally with a Dapr build from master)

File: daprdocs/content/en/python-sdk-docs/python-client.md
Comment (cicoyle) [inline daprdocs/content/en/python-sdk-docs/python-client.md:81]:
```suggestion
Example of handling `DaprGrpcError` exceptions when using the Dapr Python-SDK:
```

File: examples/error_handling/README.md
Comment (cicoyle) [inline examples/error_handling/README.md:13]:
```suggestion
## Install Dapr Python-SDK
```

File: examples/error_handling/README.md
Comment (cicoyle) [inline examples/error_handling/README.md:3]: Its a nit, but I think we want Python capitalized.
```suggestion
This guide demonstrates handling `DaprGrpcError` errors when using the Dapr Python-SDK. It's important to note that not all Dapr gRPC status errors are currently captured and transformed into a `DaprGrpcError` by the SDK. Efforts are ongoing to enhance this aspect, and contributions are welcome. For detailed information on error handling in Dapr, refer to the [official documentation](https://docs.dapr.io/reference/).
```

Comment (berndverst) [reply examples/error_handling/README.md:3]: I don't want to retrigger some of the CI tasks because of this trivial docs change. Let's do that in a follow up PR after everything is merged here.

**Pull request merged**
